content/2015.07.cliquet.rationale.rst

```diff
@@ -0,0 +1,518 @@
+
+Désormais, quand un projet souhaite bénéficier d'une API REST pour stocker et consommer
+des données, il est possible d'utiliser le **protocole HTTP** proposé
+et de se concentrer sur l'essentiel. Cela vaut aussi pour les clients, où
+la majorité du code d'interaction avec le serveur est réutilisable.
+
+.. epigraph::
+
+    Comment pouvons-nous vérifier que le service est opérationnel ? Quels indicateurs StatsD ?
+    Est-ce que Sentry est bien configuré ? Comment déployer une nouvelle version
+    sans casser les applications clientes ?
+
+Comme *Cliquet* fournit tout ce qui est nécessaire pour être conforme avec les
+exigences de la **mise en production**, le passage du prototype au service opérationnel
+est très rapide ! De base le service répondra aux attentes en terme supervision, configuration,
+déploiement et dépréciation de version. Et si celles-ci évoluent, il suffira
+de faire évoluer le toolkit.
+
+.. epigraph::
+
+    Quel backend de stockage pour des documents JSON ? Comment faire si l'équipe
+    de production impose PostgreSQL ? Et si on voulait passer à Redis ou en
+    mémoire pour lancer les tests ?
+
+En terme d'implémentation, nous avons choisi de **fournir des abstractions**.
+En effet, nous avions deux services dont le coeur consistait
+à exposer un *CRUD* en *REST*, persistant des données JSON dans un backend.
+Comme *Pyramid* et *Cornice* ne fournissent rien de tout prêt pour ça,
+nous avons voulu introduire des classes de bases pour abstraire les notions
+de resource REST et de backend de stockage.
+
+Dans le but de tout rendre optionnel et «pluggable», **tout est configurable**
+depuis le fichier ``.ini`` de l'application. Ainsi tous les projets qui utilisent
+le toolkit se déploieront de la même manière : seuls quelques éléments de configuration
+les distingueront.
+
+.. image:: {filename}/images/cliquet-notes-whiteboard.jpg
+    :alt: Une réunion à Paris...
+    :align: center
+
+
+Le protocole
+============
+
+.. epigraph::
+
+    Est-ce suffisant de parler d'«API REST» ? Est-ce bien nécessaire de
+    relire la spec HTTP à chaque fois ? Pourquoi réinventer un protocole complet
+    à chaque fois ?
+
+Quand nous développons un (micro)service Web, nous dépensons généralement beaucoup
+trop d'énergie à (re)faire des choix (arbitraires).
+
+Nous ne prendrons pas la peine de lister ici tout ce qui concerne la dimension
+de la spécification HTTP pure, qui nous impose le format des headers,
+le support de CORS, la négocation de contenus (types mime), la différence entre
+authentification et autorisation, la cohérence des code status...
+
+Les choix principaux du protocole concernent surtout :
+
+* **Les resources REST** : Les deux URLs d'une resource, pour la collection
+  et les instances, acceptent des verbes et des headers précis.
+* **Les formats** : le format et la structure JSON des réponses est imposé, ainsi
+  que la `pagination des listes <{filename}/2015.05.continuation-token.rst>`_
```

Comment:
oui c'est (la manière dont fonctionne pelican)[http://docs.getpelican.com/en/3.6.0/content.html#linking-to-internal-content]